Mount Name Filter

Marci696 opened the pull request:
<img width="531" height="360" alt="image" src="https://github.com/user-attachments/assets/6aa8bd77-a061-4121-befa-afb18d26f62e" />

Showcase: https://www.youtube.com/watch?v=vj43-XeyrAw

I am open to other suggestions on where to put it and what width the input should have.

Sorry for some reformatted code, used the Jetbrains Rider formatter which should completely follow your .editorconfig
If you want me to undo them or adjust .editorconfig, thats fine.

Marci696 committed:
first working version of mount name filter

CMDRNuffin commented on UI/MountGroupPage.cs:
```diff
@@ -240,7 +277,22 @@ string GetFastMountsText()
+                .Where((mountData) =>
+                    mountData.Name.ExtractText().Contains(_nameFilter, StringComparison.OrdinalIgnoreCase)
+                )
+                .ToList();
```
We should probably not filter the list again on _every_ frame, but instead cache the result and only recalculate if the filter changes:

```csharp
// required additional fields:
// - List<MountData>? _filteredMounts;
// - (int UnlockedCount, string Text) _lastFilter;
if (string.IsNullOrWhiteSpace(_nameFilter)) return unlockedMounts;
if (_filteredMounts is null
    || unlockedMounts.Count != _lastFilter.UnlockedCount
    || !_nameFilter.Equals(_lastFilter.Text, StringComparison.OrdinalIgnoreCase))
{
    // IMO respecting whitespace at the start and end doesn't really make sense
    // because it is rarely entered intentionally and only invites "doesn't find X" bug reports
    // because the user actually searched for " X" or "X " and the like
    string effectiveFilter = _nameFilter.Trim();
    _filteredMounts = unlockedMounts.Where(
        mountData => mountData.Name.ExtractText()
            .Contains(effectiveFilter, StringComparison.OrdinalIgnoreCase)).ToList();
}

_lastFilter = (UnlockedCount: unlockedMounts.Count, Text: _nameFilter);
return _filteredMounts;
```

As an aside: I just (several hours after hacking the above code together at work before getting sidetracked and forgetting the code review until now) noticed that my implementation has an unintentional feature... if the user deletes the entire filter text at once, we will still have a cache of the last actual filtered list, so if they then paste the text back updating the filter is essentially a no-op.

Marci696 replied:
implemented a MemoryCache with a tuple of name filter string and count of unlocked mounts as key

CMDRNuffin replied:
There's a reason I added a bunch of code to that comment. Pulling in a whole other dependency just to not filter the list every frame seems a bit overkill. Like deploying nuclear warheads against a pidgeon in your back yard. (Why'd you even go there? Was it because I said "cache"?)

The memory cache might do what you want, but it also keeps a whole bunch of garbage around that we don't care about:
- If the user unlocks a new mount, all existing cache entries are now useless because the number of unlocked mounts can never go down.
- If the user types something, we cache every. single. incomplete. version of that text, when all we likely need is the last one ("lsn" -oops- "lanner" - BAM, 8 cache entries).

Further wrinkles:
- Now you have a (admittedly probably very slight) performance penalty every frame to look up the cache entry and probably do some internal maintenance to detect stale entries instead of just comparing two values.
- If the user closes the window, waits a minute and opens the window back up, the cache assumes the entry is stale, even if neither the list of unlocked mounts nor the filter text changed.
- Having to make _everything_ disposable just so the cache in the inner innards of the config window can be cleanly cleaned up automatically.

Marci696 replied:
- the disposable part is admittadly annoying
- it doesn't keep the garbage for long

> If the user closes the window, waits a minute and opens the window back up, the cache assumes the entry is stale, even if neither the list of unlocked mounts nor the filter text changed.

that is to not keep the garbage for long. that could be prolonged but there is no real point. going through the list of 1 to 300 mounts that a user can own won't take long anyway.

> If the user unlocks a new mount, all existing cache entries are now useless because the number of unlocked mounts can never go down.

which is why they will be removed and are not just a dictionary which gets bigger and bigger

> If the user types something, we cache every. single. incomplete. version of that text, when all we likely need is the last one ("lsn" -oops- "lanner" - BAM, 8 cache entries).

that is also a feature. i can type lanner. do something. switch list, type lanner again and it is still there in cache.

> Like deploying nuclear warheads against a pidgeon in your back yard. (Why'd you even go there? Was it because I said "cache"?)

i was more thinking of using the right tool for the job that is easy to understand. juggling between 2 states of last and current is not the easyiest thing to read.

**but i can still do that with previous and current state, sure.** 
it just seemed cleaner to me with a cache key where it is clear when it is busted.. i would have preferred not to use external dependency, and i do see it is an eye sore. i do find it confusing there is so many features already included in C# but nothing of that sort?

Marci696 replied:
the cache part was more inspired by your StringCache.Mounts btw.

which does not clear after closing the config window and is not needed anywhere else by the application. so I opted for something that does clear the inputs

Marci696 replied:
having explained my reasoning.

if you are still strongly against it (what i assume you are), just say so and I change it to what your first comment was about

CMDRNuffin replied:
I am still strongly against, yes. (Not against having a dedicated cache object for that per se, just the ginormous dependency and keeping around multiple past filtered lists.)

And the `StringCache`s behave the way they do because each text the plugin uses doesn't change during the runtime of the program (at all), but they are read a _lot_. It's much nicer, performance-wise, to create the dynamic strings (mount names, page numbers etc.) _once_ and then only ever referring to them by ID than creating them over and over and filling the heap with garbage for the GC to collect every few seconds.

CMDRNuffin commented on UI/MountGroupPage.cs:
```diff
@@ -69,7 +74,12 @@ private void DrawMountGroup(MountGroup group)
+            int pages = MountRenderer.GetPageCount(filteredAndUnlockedMounts.Count);
             if (pages == 0)
             {
                 ImGui.Text("Please unlock at least one mount."u8);
```
If `unlockedMounts` contains items, this text should change to `No mounts match the current filter.` or similar.

CMDRNuffin commented on UI/MountGroupPage.cs:
```diff
@@ -141,10 +172,13 @@ private void RenderMountListPage(int page, MountGroup group, List<MountData> unl
                 () => MountRenderer.Update(
-                    _plugin.MountRegistry.GetUnlockedMounts(),
+                    // todo discuss whether it is intended, when "Select all" is used, it will only add the ones shown from filter
+                    unlockedAndFilteredMounts,
```
Only affecting the current filter is fine (in fact I'd expect it to work that way), but IMO if the list is filtered the confirmation message should say "Do you really want to (un)select all (un)selected mounts matching 'FILTERTEXT'?"

i.e. add a check and cases to the `pageInfo` calculation above that produce `all selected mounts matching '{_nameFilter}'` and `all unselected mounts matching '{_nameFilter}'`.

Marci696 replied:
<img width="583" height="103" alt="image" src="https://github.com/user-attachments/assets/9903bed5-a25d-4fcc-af99-03ada3bd06a7" /> 
<img width="531" height="102" alt="image" src="https://github.com/user-attachments/assets/49837527-6d9a-45dd-bbd8-3f45d7cf6987" />


----

although I think this sounds better and I have added it as current commit
<img width="448" height="98" alt="image" src="https://github.com/user-attachments/assets/fb1bdba1-17de-468a-919e-25f36032750f" /> 
<img width="475" height="108" alt="image" src="https://github.com/user-attachments/assets/14662c53-fa26-4fee-a2bd-cf69febead2b" />

Marci696 committed:
change deprecation error to warning. warning should be fixed in other PR.

Marci696 committed:
added caching to MountGroupPage name filter

Marci696 committed:
added different empty mounts group message when name filter is set.

Marci696 committed:
Merge branch 'CMDRNuffin:main' into add-mount-name-filter

Marci696 committed:
Merge branch 'add-mount-name-filter' of https://github.com/Marci696/fork_BetterMountRoulette into add-mount-name-filter

Marci696 committed:
changed empty message, as context to current filter is already there

Marci696 committed:
added different confirm text when name filter is set

Marci696 committed:
changed confirm prompt message to what I believe sounds better

Marci696 commented on UI/WindowManager.cs:
```diff
@@ -4,14 +4,18 @@
+    "Microsoft.Usage",
+    "CA2213:DisposableFieldsShouldBeDisposed",
+    Justification = "Plugin and services are owned/disposed externally."
+)]
```
i do find that quite ugly. it came up as error as soon as I made it disposable.

plugin is disposed by Dalamud and services are already disposed in Plugin

Marci696 commented on BetterMountRoulette.csproj:
```diff
@@ -48,10 +48,18 @@
 
+  <PropertyGroup>
+    <WarningsNotAsErrors>CS0618</WarningsNotAsErrors>
+  </PropertyGroup>
```
that deprecation error came up and didn't come up before. probably due to new dalamud release?

should be new PR to update it.

there are now warnings for

IClientState.LocalPlayer is deprecated
IClientState.LocalContentId is deprecated

CMDRNuffin replied:
I actually just updated it, so you can (and should) remove this once you've grabbed my latest version.

CMDRNuffin commented on UI/MountGroupPage.cs:
```diff
@@ -69,18 +77,27 @@ private void DrawMountGroup(MountGroup group)
+                ImGui.Text(
+                    _nameFilter.IsNullOrEmpty()
+                        ? "Please unlock at least one mount."u8
+                        : "No mounts match the current filter."u8
```
That's not what I said. If the user doesn't have any mounts unlocked, the message should be "Please unlock at least one mount", regardless of whether they are currently filtering. Please use `unlockedMounts.Count` to check whether the emptyness of the list is because of the filter or because the user doesn't have any.

Note that just because `_nameFilter` cannot be empty if the result is empty and the user owns mounts, this doesn't mean that the user owns mounts if the result is empty and `_nameFilter` is not.

Marci696 replied:
it took me quite a few times to understand the issue you see.

- the name filter is rendered even when there are 0 unlocked mounts
- user can input something in that filter even when there are 0 unlocked mounts
- empty message is then about the filter and not about the fact the user has 0 mounts.

are you then fine with

```c#
           List<MountData> filteredAndUnlockedMounts = GetFilteredMounts(unlockedMounts);

            int pages = MountRenderer.GetPageCount(filteredAndUnlockedMounts.Count);
            if (pages == 0)
            {
                ImGui.Text(
                    unlockedMounts.Count == 0
                        ? "Please unlock at least one mount."u8
                        : "No mounts match the current filter."u8
                );
            }
```

 or have i still misunderstood or missed something?

Marci696 replied:
can also go further and not even render the filter bar when there are no mounts

```c#
            if (unlockedMounts.Count > 0)
            {
                DrawNameFilter();
            }

            List<MountData> filteredAndUnlockedMounts = GetFilteredMounts(unlockedMounts);

            int pages = MountRenderer.GetPageCount(filteredAndUnlockedMounts.Count);
            if (pages == 0)
            {
                ImGui.Text(
                    unlockedMounts.Count == 0
                        ? "Please unlock at least one mount."u8
                        : "No mounts match the current filter."u8
                );
            }
```

<img width="542" height="180" alt="image" src="https://github.com/user-attachments/assets/a050684b-35ca-4261-86c1-7ba8e2839f75" />

CMDRNuffin replied:
Yes, you understood correctly. Also, I like your idea of not having the filter at all when there are no mounts to filter, feel free to add it to the PR.

Marci696 committed:
- don't draw name filter when there are no mounts unlocked
- make sure message for no unlocked mounts is displayed independent of name filter string

Marci696 committed:
Merge remote-tracking branch 'origin/main' into add-mount-name-filter

# Conflicts:
#	BetterMountRoulette.csproj
#	packages.lock.json

Marci696 committed:
got rid of Microsoft Memory cache and replaced the filter logic with proposed solution from PR

Marci696 commented on UI/MountGroupPage.cs:
```diff
@@ -240,7 +293,33 @@ string GetFastMountsText()
+                : FastMode.Off;
+    }
+
+    private List<MountData> ApplyFilterAndGetFilteredMounts(List<MountData> unlockedMounts)
```
function name does not sound great, but I didn't want to keep "GetFilteredMounts" when the function has a side effect of setting _lastFilter

CMDRNuffin replied:
IMO it would have been fine, since setting `_lastFilter` and `_filteredMounts` is an internal optimization that's not particularly relevant to a potential reader of the call site. Then again it's not as if having that information hurts anyone so whatever ¯\\\_(ツ)\_/¯

Marci696 reviewed (commented):
i have now implemented your proposed solution with _lastFilter and removed the extra MemoryCache dependency + dispose() logic.

the only thing I belive to have changed from your proposed solution is the _nameFilter.trim() part, as I am making sure _nameFilter is always trimmed beforehand with

```c#
        if (ImGui.InputTextWithHint("###nameFilter"u8, "Search for name..."u8, ref _nameFilter))
        {
            _nameFilter = _nameFilter.Trim();
        }
```

CMDRNuffin commented on UI/MountGroupPage.cs:
```diff
@@ -112,11 +132,35 @@ private void DrawMountGroup(MountGroup group)
+        // as _nameFilter is directly changed per reference.
+        if (ImGui.InputTextWithHint("###nameFilter"u8, "Search for name..."u8, ref _nameFilter))
+        {
+            _nameFilter = _nameFilter.Trim();
```
Did you run this and actually enter a few spaces? Because as far as I can tell, they would be stripped from the input while the user is still typing (which is not conductive to a good user experience). That's why I didn't do it this way in my example for how to do the caching – I figured filtering the list once more with the same effective value when the user entered a space was good enough.

If you're really concerned about only filtering when the effective filter text changes (which is a weird hill to die on, but ok 😄), what you could do instead is this: Have a second `_rawNameFilter` field that's used for storing the text thats put in by and displayed to the user. Update `_nameFilter` to the effective version whenever the raw version changes.

Marci696 replied:
yes it works because ImGui.Input() has its own independent state while the input is focused.

it updates the passed reference and reads it while the input is not focused. but while the input is focused, it uses its own independent state to display the current input value.

so while the current thing might look weird and incorrect, it is actually quite simple as long as it is understood what ImGui does here.

https://www.youtube.com/watch?v=DzUq0hhZSfA
